gptme/codeblock.py

```diff
@@ -0,0 +1,84 @@
+from collections.abc import Generator
+from dataclasses import dataclass
+from xml.etree import ElementTree
+
+
+@dataclass
+class Codeblock:
+    lang: str
+    content: str
+    path: str | None = None
+
+    # init path in __post_init__ if path is None and lang is pathy
+    def __post_init__(self):
+        if self.path is None and self.is_filename:
+            self.path = self.lang
+
+    def to_markdown(self) -> str:
+        return f"```{self.lang}\n{self.content}\n```"
+
+    def to_xml(self) -> str:
+        return f'<codeblock lang="{self.lang}" path="{self.path}">\n{self.content}\n</codeblock>'
+
+    @classmethod
+    def from_markdown(cls, content: str) -> "Codeblock":
+        if content.strip().startswith("```"):
+            content = content[3:]
+        if content.strip().endswith("```"):
+            content = content[:-3]
+        lang = content.splitlines()[0].strip()
+        return cls(lang, content[len(lang) :])
+
+    @classmethod
+    def from_xml(cls, content: str) -> "Codeblock":
+        """
+        Example:
+          <codeblock lang="python" path="example.py">
+          print("Hello, world!")
+          </codeblock>
+        """
+        root = ElementTree.fromstring(content)
+        return cls(root.attrib["lang"], root.text or "", root.attrib.get("path"))
+
+    @property
+    def is_filename(self) -> bool:
+        return "." in self.lang or "/" in self.lang
+
+    @classmethod
+    def iter_from_markdown(cls, markdown: str) -> list["Codeblock"]:
+        return list(_extract_codeblocks(markdown))
+
+
+def _extract_codeblocks(markdown: str) -> Generator[Codeblock, None, None]:
+    # speed check (early exit): check if message contains a code block
+    backtick_count = markdown.count("```")
+    if backtick_count < 2:
+        return
+    # check that end tick has a newline (no language tag)
+    # if not markdown.count("```\n"):
+    #     return
```

Comment:
```suggestion
    # check that end tick has a newline (no language tag)
    # if not markdown.count("```\n"):
    #     return
```